domain/model/service/service.go

```diff
@@ -709,3 +730,60 @@ func (s WatchableService) WatchModel(ctx context.Context, modelUUID coremodel.UU
 		}),
 	)
 }
+
+// WatchModelCloudCredential returns a new NotifyWatcher watching for changes that
+// result in the cloud spec for a model changing. The changes watched for are:
+// - updates to model cloud.
+// - updates to model credential.
+// - changes to the credential set on a model.
+// The following errors can be expected:
+// - [modelerrors.NotFound] when the model is not found.
+func (s *WatchableService) WatchModelCloudCredential(ctx context.Context, modelUUID coremodel.UUID) (watcher.NotifyWatcher, error) {
+	if err := modelUUID.Validate(); err != nil {
+		return nil, errors.Errorf("invalid model UUID watching model cloud credential: %w", err)
+	}
+
+	// Get the model's cloud and credential UUID to use in the watcher filters.
+	cloudUUID, credentialUUID, err := s.st.GetModelCloudAndCredential(ctx, modelUUID)
+	if err != nil {
+		return nil, errors.Errorf("getting model cloud and credential: %w", err)
+	}
+
+	var currentCredentialUUID atomic.Pointer[credential.UUID]
+	currentCredentialUUID.Store(&credentialUUID)
+
+	// The mapper is used to filter out any model events where the credential UUID has not changed.
+	mapper := func(ctx context.Context, txnRunner database.TxnRunner, events []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
+		// Get the current model credential UUID.
+		_, newCredentialUUID, err := s.st.GetModelCloudAndCredential(ctx, modelUUID)
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		gotNewCredential := *currentCredentialUUID.Load() != newCredentialUUID
+		var out []changestream.ChangeEvent
+		for _, e := range events {
+			if e.Namespace() != "model" || gotNewCredential {
+				out = append(out, e)
+			}
+		}
+		currentCredentialUUID.Store(&newCredentialUUID)
+		return out, nil
+	}
+
+	result, err := s.watcherFactory.NewNotifyMapperWatcher(
+		mapper,
+		eventsource.PredicateFilter("model", changestream.Changed, func(s string) bool {
+			return s == modelUUID.String()
+		}),
+		eventsource.PredicateFilter("cloud", changestream.Changed, func(s string) bool {
+			return s == cloudUUID.String()
+		}),
+		eventsource.PredicateFilter("cloud_credential", changestream.Changed, func(s string) bool {
+			return s == currentCredentialUUID.Load().String()
```

Comment:
Hi Ian, not sure if this could cause an unintended filtering due to a data race? From my understanding, we subscribe to db changes based on filter, and then write to currentCredentialUUID with mapper if there's any changes in credential UUID.

Logic flow:
1. Filter -> Filters using the current credential UUID
2. Mapper -> Updates current credential UUID to new credential UUID

But not sure if the following case could happen?
1. Filter1 -> Sees old credential uuid
2. Filter2 -> Sees old credential uuid
3. Mapper1 -> Updates to new credential uuid
4. Mapper2 -> Skipped due to Filter 2

Suggestion:
Would it be safer to drop the predicateFilter for currentCredentialUUID if we are checking it in the mapper itself?

Reply:
To be safe, we don't lose much by use a namespace filter so have updated.